src/v/datalake/coordinator/frontend.cc

```diff
@@ -0,0 +1,259 @@
+    return _connection_cache->local()
+      .with_node_client<proto_t>(
+        _self,
+        ss::this_shard_id(),
+        leader_id,
+        rpc_timeout,
+        [request = std::move(request)](proto_t proto) mutable {
+            return (proto.*Func)(
+              std::move(request),
+              ::rpc::client_opts{model::timeout_clock::now() + rpc_timeout});
+        })
+      .then(&::rpc::get_ctx_data<resp_t>)
+      .then([leader_id](result<resp_t> r) {
+          if (r.has_error()) {
+              vlog(
+                datalake::datalake_log.warn,
+                "got error {} on coordinator {}",
+                r.error().message(),
+                leader_id);
+              return resp_t{datalake::coordinator_errc::timeout};
+          }
+          return r.value();
+      });
+}
+
+template<auto LocalFunc, auto RemoteFunc, typename req_t>
+requires requires(
+  datalake::coordinator::frontend f, const model::ntp& ntp, req_t req) {
+    (f.*LocalFunc)(std::move(req), ntp, ss::shard_id{0});
+}
+auto frontend::process(req_t req, bool local_only) {
+    using resp_t = req_t::resp_t;
+    return ensure_topic_exists().then([req = std::move(req), local_only, this](
+                                        bool exists) mutable {
+        if (!exists) {
+            return ss::make_ready_future<resp_t>(
+              resp_t{datalake::coordinator_errc::coordinator_topic_not_exists});
+        }
+        auto cp = coordinator_partition(req.topic_partition());
+        model::ntp c_ntp{
+          model::datalake_coordinator_nt.ns,
+          model::datalake_coordinator_nt.tp,
+          cp.value()};
+        auto leader = _leaders->local().get_leader(c_ntp);
+        if (leader == _self) {
+            auto shard = _shard_table->local().shard_for(c_ntp);
+            if (shard) {
+                return (this->*LocalFunc)(
+                  std::move(req), std::move(c_ntp), shard.value());
+            }
+        } else if (leader && !local_only) {
+            return remote_dispatch<RemoteFunc>(std::move(req), leader.value());
+        }
+        return ss::make_ready_future<resp_t>(
+          resp_t{datalake::coordinator_errc::not_leader});
+    });
+}
+
+// -- explicit instantiations ---
+template auto
+  frontend::remote_dispatch<&frontend::client::add_translated_data_files>(
+    add_translated_data_files_request, model::node_id);
+
+template auto frontend::process<
+  &frontend::add_translated_data_files_locally,
+  &frontend::client::add_translated_data_files>(
+  add_translated_data_files_request, bool);
+
+template auto
+  frontend::remote_dispatch<&frontend::client::fetch_latest_data_file>(
+    fetch_latest_data_file_request, model::node_id);
+
+template auto frontend::process<
+  &frontend::fetch_latest_data_file_locally,
+  &frontend::client::fetch_latest_data_file>(
+  fetch_latest_data_file_request, bool);
+
+// -- explicit instantiations ---
+
+frontend::frontend(
+  model::node_id self,
```

Comment:
Not really a suggestion, but here's another instance where maybe making leader_router slightly more generic would be useful (or replacing it with this new code)

Reply:
ya I looked at leader_router, felt like it is a bit more verbose (and general purpose) than this, but agree these two approaches can be merged somehow.